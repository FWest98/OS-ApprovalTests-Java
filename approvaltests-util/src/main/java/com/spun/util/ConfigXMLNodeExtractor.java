package com.spun.util;

import com.spun.util.io.XMLNodeExtractor;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import java.util.ArrayList;
import java.util.HashMap;

public class ConfigXMLNodeExtractor implements XMLNodeExtractor
{
  public void extractProperty(Node node, HashMap<String, Object> properties)
  {
    //String name = node.getNodeName();
    if (XMLNodeExtractor.Utils.extractSingleton(node, properties))
    {
      // do nothing
    }
    else if (node.getChildNodes().getLength() > 0)
    {
      NodeList childNodes = node.getChildNodes();
      ArrayList<String> v = new ArrayList<String>();
      for (int i = 0; i < childNodes.getLength(); i++)
      {
        if (childNodes.item(i).getNodeName().equals("VALUE"))
        {
          v.add(StringUtils.loadNullableString(childNodes.item(i).getChildNodes().item(0).getNodeValue()));
        }
      }
      properties.put(node.getNodeName(), StringUtils.toArray(v));
    }
  }
}
