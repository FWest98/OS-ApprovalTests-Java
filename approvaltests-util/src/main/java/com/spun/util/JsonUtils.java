package com.spun.util;

import com.google.gson.*;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;
import org.lambda.functions.Function1;

import java.io.IOException;
import java.time.Instant;

public class JsonUtils
{
  public static String prettyPrint(String json)
  {
    if (!ObjectUtils.isClassPresent("com.google.gson.Gson"))
    {
      throw new RuntimeException(
          "Missing Gson dependency\n  Pretty print uses Gson parser.\n  You can get this from the maven repo \n  or https://github.com/google/gson");
    }
    try
    {
      Gson gson = new GsonBuilder().setPrettyPrinting().create();
      JsonElement je = JsonParser.parseString(json);
      return gson.toJson(je);
    }
    catch (JsonSyntaxException e)
    {
      return String.format("Error:%s\nJson:\n%s", e.getMessage(), json);
    }
  }
  public static String asJson(Object o)
  {
    return asJson(o, (b) -> b);
  }
  public static <T> String asJsonWithBuilder(Object o, Function1<T, T> gsonBuilder, Class<T> gsonBuilderClass)
  {
    if (!ObjectUtils.isThisInstanceOfThat(gsonBuilderClass, com.google.gson.GsonBuilder.class))
    { throw new FormattedException("Class must be of type %s", GsonBuilder.class.getName()); }
    return asJson(o, (Function1<GsonBuilder, GsonBuilder>) gsonBuilder);
  }
  public static <T> String asJson(Object o, Function1<GsonBuilder, GsonBuilder> gsonBuilder)
  {
    GsonBuilder builder = new GsonBuilder().setPrettyPrinting();
    builder = addHandlingForDateObjects(builder);
    Gson gson = gsonBuilder.call(builder).create();
    return gson.toJson(o);
  }
  private static GsonBuilder addHandlingForDateObjects(GsonBuilder builder)
  {
    builder = builder.registerTypeAdapter(Instant.class, new InstantAdapter());
    return builder;
  }
  public static class InstantAdapter extends TypeAdapter<Instant>
  {
    @Override
    public void write(JsonWriter jsonWriter, Instant instant) throws IOException
    {
      jsonWriter.value(instant.toString());
    }
    @Override
    public Instant read(JsonReader jsonReader) throws IOException
    {
      throw new IOException("Never called");
    }
  }
}
