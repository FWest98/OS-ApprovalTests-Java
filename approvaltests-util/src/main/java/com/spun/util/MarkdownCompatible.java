package com.spun.util;

public interface MarkdownCompatible
{
  String toMarkdown();
}
