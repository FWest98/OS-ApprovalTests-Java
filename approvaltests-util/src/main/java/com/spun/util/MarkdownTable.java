package com.spun.util;

import org.lambda.functions.Function1;
import org.lambda.query.Queryable;

public class MarkdownTable implements MarkdownCompatible
{
  protected String markdown;
  public static <I, O> MarkdownTable create(I[] inputs, Function1<I, O> o, String column1, String column2)
  {
    MarkdownTable table = new MarkdownTable().withColumnHeaders(column1, column2);
    for (I input : inputs)
    {
      table.addRow(input, o.call(input));
    }
    return table;
  }
  public static MarkdownTable withHeaders(String... columnNames)
  {
    MarkdownTable table = new MarkdownTable();
    return table.withColumnHeaders(columnNames);
  }
  public <I> MarkdownTable addRowsForInputs(I[] inputs, Function1<I, Object>... transfers)
  {
    for (I input : inputs)
    {
      Queryable<Object> row = Queryable.as(transfers).select(f -> f.call(input));
      row.add(0, input);
      addRow(row.toArray());
    }
    return this;
  }
  public MarkdownTable addRow(Object... columns)
  {
    markdown += printRow(columns);
    return this;
  }
  public MarkdownTable withColumnHeaders(String... headers)
  {
    markdown = printColumnHeaders(headers);
    return this;
  }
  @Override
  public String toString()
  {
    return toMarkdown();
  }
  @Override
  public String toMarkdown()
  {
    return markdown;
  }
  public static String printColumnHeaders(String... headers)
  {
    return printRow(headers) + printRow(ArrayUtils.of("---", headers.length));
  }
  public static String printRow(Object... columns)
  {
    StringBuffer b = new StringBuffer();
    b.append("|");
    for (int x = 0; x < columns.length; ++x)
    {
      b.append(String.format(" %s |", columns[x]));
    }
    b.append("\n");
    return b.toString();
  }
}
