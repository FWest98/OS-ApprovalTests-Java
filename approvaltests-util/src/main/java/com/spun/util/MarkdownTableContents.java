package com.spun.util;

public class MarkdownTableContents implements MarkdownTableElement, Resizable
{
  private String contents;
  private int    padUntil;
  public MarkdownTableContents(String contents)
  {
    this.contents = contents;
  }
  @Override
  public String toString()
  {
    return " " + StringUtils.pad(contents, padUntil) + " ";
  }
  public int getLength()
  {
    return contents.length();
  }
  public void setPadding(int length)
  {
    padUntil = length;
  }
}
