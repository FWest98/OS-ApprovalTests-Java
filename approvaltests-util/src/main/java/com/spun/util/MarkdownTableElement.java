package com.spun.util;

public interface MarkdownTableElement {
    MarkdownTableElement DELIMITER = new MarkdownTableElement() {
        @Override
        public String toString() {
            return "|";
        }
    };
    MarkdownTableElement NEWLINE = new MarkdownTableElement() {
        @Override
        public String toString() {
            return "\n";
        }
    };
}
