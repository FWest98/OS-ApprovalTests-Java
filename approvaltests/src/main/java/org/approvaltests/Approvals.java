package org.approvaltests;

import com.spun.util.*;
import com.spun.util.persistence.ExecutableQuery;
import com.spun.util.persistence.Loader;
import com.spun.util.persistence.SqlLoader;
import org.approvaltests.approvers.ApprovalApprover;
import org.approvaltests.approvers.FileApprover;
import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.core.ApprovalWriter;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.StackTraceNamer;
import org.approvaltests.reporters.ExecutableQueryFailure;
import org.approvaltests.writers.*;
import org.lambda.actions.Action0;
import org.lambda.functions.Function1;
import org.lambda.query.Query;

import java.io.File;
import java.io.FileFilter;
import java.io.FilenameFilter;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

public class Approvals
{
  public static Loader<ApprovalNamer> namerCreater = new Loader<ApprovalNamer>()
  {
    public ApprovalNamer load()
    {
      return new StackTraceNamer();
    }
  };
  public static void verify(String response)
  {
    verify(new ApprovalTextWriter(response, "txt"));
  }
  public static void verify(Object o)
  {
    verify("" + o);
  }
  public static <T> void verifyAll(String label, T[] array)
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"));
  }
  public static <T> void verifyAll(String header, String label, T[] array)
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"));
  }
  public static <T> void verifyAll(T[] values, Function1<T, String> f1)
  {
    String text = ArrayUtils.toString(values, f1);
    verify(new ApprovalTextWriter(text, "txt"));
  }
  public static <T> void verifyAll(String header, T[] values, Function1<T, String> f1)
  {
    verifyAll(header, Arrays.asList(values), f1);
  }
  public static <T> void verifyAll(String header, Iterable<T> array, Function1<T, String> f1)
  {
    String text = formatHeader(header) + ArrayUtils.toString(array, f1);
    verify(new ApprovalTextWriter(text, "txt"));
  }
  private static String formatHeader(String header)
  {
    return StringUtils.isEmpty(header) ? "" : header + "\n\n\n";
  }
  public static <T> void verifyAll(String label, Iterable<T> array)
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"));
  }
  public static <T> void verifyAll(String header, String label, Iterable<T> array)
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"));
  }

  public static void verifyHtml(String response)
  {
    verify(new ApprovalTextWriter(response, "html"));
  }
  public static void verify(File generateFile)
  {
    verify(new FileApprovalWriter(generateFile));
  }

  public static void verify(ApprovalWriter writer, ApprovalNamer namer, ApprovalFailureReporter reporter)
  {
    verify(new FileApprover(writer, namer), reporter);
  }
  public static void verify(ApprovalWriter writer)
  {
    verify(writer, createApprovalNamer(), getReporter());
  }
  public static void verifyXml(String xml)
  {
    verify(new ApprovalXmlWriter(xml));
  }
  public static void verify(ApprovalApprover approver)
  {
    verify(approver, getReporter());
  }
  public static void verify(ApprovalApprover approver, ApprovalFailureReporter reporter)
  {
    if (!approver.approve())
    {
      approver.reportFailure(reporter);
      approver.fail();
    }
    else
    {
      approver.cleanUpAfterSuccess(reporter);
    }
  }
  public static void verify(ExecutableQuery query)
  {
    verify(new ApprovalTextWriter(query.getQuery(), "txt"), createApprovalNamer(),
        new ExecutableQueryFailure(query));
  }
  public static void verify(Map<?, ?> map)
  {
    verify(new ApprovalTextWriter(StringUtils.toString(map), "txt"));
  }
  public static void verify(ResultSet rs)
  {
    verify(new ResultSetApprovalWriter(rs));
  }
  public static <T> void verify(SqlLoader<T> loader)
  {
    verify(new SqlLoader.ExecutableWrapper<T>(loader));
  }
  public static ApprovalNamer createApprovalNamer()
  {
    return namerCreater.load();
  }

  public static void verifyEachFileInDirectory(File directory)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles());
  }
  public static void verifyEachFileInDirectory(File directory, FileFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  public static void verifyEachFileInDirectory(File directory, FilenameFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  private static void verifyEachFileAgainstMasterDirectory(File[] files) throws Error
  {
    ApprovalNamer namer = createApprovalNamer();
    String dirName = namer.getSourceFilePath() + File.separator + namer.getApprovalName() + ".Files";
    File approvedDirectory = new File(dirName);
    List<File> mismatched = new ArrayList<File>();
    for (File f : files)
    {
      if (!f.isDirectory())
      {
        try
        {
          verify(new DirectoryToDirectoryWriter(f, approvedDirectory));
        }
        catch (Throwable e)
        {
          mismatched.add(f);
        }
      }
    }
    if (!mismatched.isEmpty())
    {
      String message = "The Following Files did not match up: " + getFileNameList(mismatched);
      throw new Error(message);
    }
  }
  private static String getFileNameList(List<File> mismatched)
  {
    return Query.select(mismatched, a -> a.getName()).toString();
  }
  public static void verifyJson(String json)
  {
    verify(JsonUtils.prettyPrint(json), "json");
  }
  public static void verify(Object actual, String fileExtensionWithoutDot)
  {
    verify(new ApprovalTextWriter("" + actual, fileExtensionWithoutDot));
  }
  public static ApprovalFailureReporter getReporter()
  {
    return ReporterFactory.get();
  }
  public static void verifyAsJson(Object o)
  {
    verify(JsonUtils.asJson(o), "json");
  }
  public static void verifyException(Action0 runnableBlock)
  {
    Throwable t = ObjectUtils.captureException(runnableBlock);
    if (t == null) { throw new FormattedException("No exception thrown when running %s", runnableBlock); }
    Approvals.verify(String.format("%s: %s", t.getClass().getName(), t.getMessage()));
  }
}