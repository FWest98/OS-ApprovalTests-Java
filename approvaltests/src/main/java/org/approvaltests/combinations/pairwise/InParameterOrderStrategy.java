package org.approvaltests.combinations.pairwise;

import com.spun.util.Tuple;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public final class InParameterOrderStrategy
{
  public static List<List<Case>> generatePairs(List<OptionsForAParameter<?>> parameters)
  {
    final List<OptionsForAParameter> accumulator = new ArrayList<>();
    Stream<OptionsForAParameter<?>> sortedBySize = parameters.stream()
        .sorted((o1, o2) -> Integer.compare(o2.size(), o1.size()));
    Stream<ArrayList<OptionsForAParameter>> arrayListStream = sortedBySize.map(parameter -> {
      accumulator.add(parameter);
      return new ArrayList<>(accumulator);
    });
    return arrayListStream.map(chunk -> crossJoin(chunk)).collect(Collectors.toList());
  }
  public static List<Case> horizontalGrowth(List<Case> cases, List<Case> pairs)
  {
    List<Case> result = new ArrayList<>();
    for (Case aCase : cases)
    {
      Tuple<String, Object> t = findMostUsedLastParameter(pairs, aCase);
      if (t.getSecond() != null)
      {
        aCase.put(t.getFirst(), t.getSecond());
      }
      pairs.removeIf(c -> c.matches(aCase));
      result.add(aCase);
    }
    return result;
  }
  public static List<Case> verticalGrowth(List<Case> pairs)
  {
    return removeDuplicates(pairs);
  }
  public static List<Case> removeDuplicates(List<Case> pairs)
  {
    List<Case> collected = new ArrayList<>();
    for (Case aCase : pairs)
    {
      if (!collected.contains(aCase))
      {
        collected.add(aCase);
      }
    }
    return collected;
  }
  public static List<Case> crossJoin(List<OptionsForAParameter> chunk)
  {
    final OptionsForAParameter multiplier = chunk.get(chunk.size() - 1);
    return new ArrayList<Case>()
    {
      {
        IntStream.range(0, chunk.get(chunk.size() - 1).size()).forEach(last -> IntStream.range(0, chunk.size() - 1)
            .forEach(column -> IntStream.range(0, chunk.get(column).size()).forEach(cursor -> this.add(new Case()
            {
              {
                // we believe this creates problematic order but we're unsure.
                OptionsForAParameter parameter = chunk.get(column);
                this.put(parameter.getPosition(), parameter.get(cursor));
                this.put(multiplier.getPosition(), multiplier.get(last));
              }
            }))));
      }
    };
  }

  private static Tuple<String, Object> findMostUsedLastParameter(List<Case> pairs, Case aCaseParameter) {
    String key = null;
    Map<Object, Integer> lastKeyCounts = new HashMap<>();
    int amount = 0;
    Object obj = null;
    for (Case aCase : pairs)
    {
      if (aCaseParameter.matches(aCase))
      {
        if (key == null)
        {
          key = aCase.getLastKey();
        }
        Object value = aCase.get(key);
        Integer count = lastKeyCounts.computeIfAbsent(value, x -> 0) + 1;
        lastKeyCounts.put(value, count);
      }
    }
    for (Object o : lastKeyCounts.keySet())
    {
      int size = lastKeyCounts.get(o);
      if (amount < size)
      {
        obj = o;
        amount = size;
      }
    }
    Tuple<String, Object> t = new Tuple<>(key, obj);
    return t;
  }

  public static List<Case> combineAppleSauce(List<Case> createManyCases, List<Case> cases)
  {
    List<Case> horizontalAndVerticalGrowth = horizontalGrowth(createManyCases, cases);
    horizontalAndVerticalGrowth.addAll(verticalGrowth(cases));
    return horizontalAndVerticalGrowth;
  }
}
