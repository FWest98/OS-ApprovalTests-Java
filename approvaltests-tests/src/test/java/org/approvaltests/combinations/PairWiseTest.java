package org.approvaltests.combinations;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import org.approvaltests.Approvals;
import org.approvaltests.combinations.pairwise.Case;
import org.approvaltests.combinations.pairwise.InParameterOrderStrategy;
import org.approvaltests.combinations.pairwise.Parameter;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import com.github.larseckart.tcr.TestCommitRevertMainExtension;

@ExtendWith(TestCommitRevertMainExtension.class)
public class PairWiseTest
{
  @Test
  public void testPairs()
  {
    PairWiseApprovals.verifyBestCoveringPairs((a, b, c, d) -> "", new Integer[]{1, 2, 3, 4, 5},
        new String[]{"a", "b", "c", "d"}, new String[]{"L", "M", "N", "O", "P"}, new Double[]{1.1, 2.2, 3.3, 4.4});
  }
  @Test
  public void testStrategyGeneratePairs()
  {
    ArrayList list = new ArrayList();
    list.add(new Parameter<>(1, new Integer[]{1, 2, 3}));
    list.add(new Parameter<>(2, new String[]{"a", "b"}));
    list.add(new Parameter<>(3, new String[]{"L", "M"}));
    list.add(new Parameter<>(4, new Double[]{1.1, 2.2, 3.3}));
    List result = InParameterOrderStrategy.generatePairs(list);
    Approvals.verifyAll("Generate Pairs", result);
  }
  @Test
  public void testRemoveDuplicateCases()
  {
    List<Case> cases = new ArrayList<>();
    cases.add(new Case(1, "Fred", false));
    cases.add(new Case(2, "Sam", true));
    cases.add(new Case(1, "Fred", false));
    String output = String.format("given cases \n%s\n", toString(cases));
    List<Case> result = InParameterOrderStrategy.verticalGrowth(cases);
    output += "\nvertical Growth results in\n" + toString(result);
    Approvals.verify(output);
  }

  public String toString(List<Case> result) {
    return result.stream().map(Case::toString).collect(Collectors.joining("\n"));
  }
}
