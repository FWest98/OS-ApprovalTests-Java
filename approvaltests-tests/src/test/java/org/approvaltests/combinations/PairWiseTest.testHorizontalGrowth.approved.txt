Cases before: 
{1=Lars, 2=1, 3=Tartu}
{1=Lars, 2=1, 3=Tallinn}
{1=Lars, 2=2, 3=Tartu}
{1=Lars, 2=2, 3=Tallinn}
{1=Lars, 2=3, 3=Tartu}
{1=Lars, 2=3, 3=Tallinn} 
 Cases after: 
{1=Lars, 2=1, 3=Tartu}
{1=Lars, 2=1, 3=Tallinn}
{1=Lars, 2=2, 3=Tartu}
{1=Lars, 2=2, 3=Tallinn}
{1=Lars, 2=3, 3=Tartu}
{1=Lars, 2=3, 3=Tallinn}
Pairs before: 
{1=Lars, 2=1, 3=Tartu}
{1=Lars, 2=1, 3=PÃ¤rnu} 
 Pairs after: 
{1=Lars, 2=1, 3=PÃ¤rnu}
 Result: 
{1=Lars, 2=1, 3=Tartu}
{1=Lars, 2=1, 3=Tallinn}
{1=Lars, 2=2, 3=Tartu}
{1=Lars, 2=2, 3=Tallinn}
{1=Lars, 2=3, 3=Tartu}
{1=Lars, 2=3, 3=Tallinn}